File: common/src/main/java/cloudcmd/common/adapters/Adapter.java
package cloudcmd.common.adapters;

import org.jboss.netty.buffer.ChannelBuffer;

import java.io.InputStream;
import java.net.URI;
import java.util.Set;

public abstract class Adapter
{
  public String ConfigDir;
  public String Type;
  public Set<String> Tags;
  public URI URI;
  public Integer Tier;
  protected boolean _isOnline;

  public Adapter()
  {
  }

  public boolean IsOnLine()
  {
    return true;
  }

  public boolean IsFull()
  {
    return false;
  }

  public void init(String configDir, Integer tier, String type, Set<String> tags, URI uri) throws Exception
  {
    ConfigDir = configDir;
    Tier = tier;
    Type = type;
    Tags = tags;
    URI = uri;
  }

  public abstract void shutdown() throws Exception;

  public boolean accepts(Set<String> tags)
  {
    if (Tags == null || Tags.size() == 0) return true;

    for (String tag : tags)
    {
      if (Tags.contains(tag)) return true;
    }

    return false;
  }

  public abstract boolean remove(String hash) throws Exception;

  public abstract boolean verify(String hash) throws Exception;
  
  public abstract void refreshCache() throws Exception;

  public abstract boolean contains(String hash) throws Exception;

  public abstract void store(InputStream data, String hash) throws Exception;

  public abstract InputStream load(String hash) throws Exception;

  public abstract ChannelBuffer loadChannel(String hash) throws Exception;

  public abstract Set<String> describe() throws Exception;
}

File: common/src/main/scala/cloudcmd/common/adapters/Adapter.scala
package cloudcmd.common.adapters

import java.net.URI
import cloudcmd.common.{BlockContext, ContentAddressableStorage}

trait Adapter extends ContentAddressableStorage {

  var ConfigDir: String = null
  var Type: String = null
  var URI: URI = null
  var Tier: Int = 0

  protected var _isOnline = false

  def IsOnLine(): Boolean = _isOnline
  def IsFull(): Boolean = false

  def init(configDir: String, tier: Int, adapterType: String, acceptsTags: Set[String], uri: URI) {
    ConfigDir = configDir
    Tier = tier
    Type = adapterType
    URI = uri

    AcceptsTags = acceptsTags
    KeepTags = acceptsTags.filterNot(_.startsWith("-"))
    IgnoreTags = (acceptsTags -- KeepTags).map(_.substring(1))
  }

  def shutdown()

  var AcceptsTags: Set[String] = null
  protected var KeepTags: Set[String] = Set()
  protected var IgnoreTags: Set[String] = Set()

  def accepts(ctx: BlockContext): Boolean = {
    if (IgnoreTags.intersect(ctx.routingTags).size > 0) return false
    if (KeepTags.size == 0) return true
    KeepTags.intersect(ctx.routingTags).size > 0
  }

  override def hashCode: Int = {
    URI.hashCode()
  }

  override def equals(obj: Any): Boolean = {
    obj.isInstanceOf[Adapter] && obj.asInstanceOf[Adapter].URI.equals(URI)
  }
}